Workaround for the 64 characters problem

vvraskin opened the pull request:
Hello,

(Copied from PR #196)

this PR addresses the issue #129. When the VM name is equal to 64 characters, the deployment is failing due to ssh problem (agent could not copy files to the VSI).

When the name with 64 characters was identified, a padding "-1" is appended to it.
I have removed the previous workaround for > 63 characters, since longer names are accepted.

@mattcui @maximilien Could you please have a look at this PR?

maximilien commented on src/bosh-softlayer-cpi/action/create_vm_test.go:
```diff
@@ -78,6 +79,8 @@ var _ = Describe("CreateVM", func() {
 				fakeStemcellFinder.FindByIdReturns(fakeStemcell, nil)
 				fakeCreatorProvider.GetReturns(fakeVmCreator)
 				fakeVmCreator.CreateReturns(fakeVm, nil)
+				Expect(helper.LengthOfHostName).ToNot(Equal(64))
```
I would like to see a test that shows the length of VMs to be less than 64 since the point is not to be longer.

maximilien commented on src/bosh-softlayer-cpi/action/create_vm_test.go:
```diff
@@ -11,6 +11,7 @@ import (
 
 	. "bosh-softlayer-cpi/softlayer/common"
 	fakescommon "bosh-softlayer-cpi/softlayer/common/fakes"
```
What about adding some tests showing failure paths?

vvraskin replied:
I've created a separate test for the case with 64 character long name, hope it makes it clearer. That's the only 'bad' case we could have here.

vvraskin replied:
We only have to consider the case when the length is _exactly_ 64 characters long, VMs got created fine when its either > or < 64.

maximilien commented:
Hi, @vvraskin,

Thanks for resubmitting. I appreciate. Made a few comments, please address.

Also, I am requesting that @jianqiu reviews this PR as well. And also @gu-bin and @zhanggbj to make sure it does not break pipeline and our acceptance tests.

Finally, note that the Beijing team is on vacation for Chinese New Year, so not sure when they will have a chance to address these.

Best,

Max

vvraskin commented:
Hello guys, as discussed with @mattcui, a friendly reminder :)

mattcui commented:
@vvraskin I just had a quick review and found a problem there, I think there doesn't need the variable [LengthOfHostName](https://github.com/cloudfoundry/bosh-softlayer-cpi-release/blob/master/src/bosh-softlayer-cpi/softlayer/common/helper/softlayer_helper.go#L20) any more, so better to remove this variable and the corresponding UT in [fs_agent_env_service_test.go](https://github.com/cloudfoundry/bosh-softlayer-cpi-release/blob/master/src/bosh-softlayer-cpi/softlayer/common/fs_agent_env_service_test.go#L111-L122), and there doesn't need [this line](https://github.com/vvraskin/bosh-softlayer-cpi/blob/0fddfc4e628fa1ae3591dc68e22ec320dbf1ef06/src/bosh-softlayer-cpi/action/create_vm.go#L101) also. What do you think? Feel free to discuss with me either here or in Slack. Thanks

vvraskin commented:
@mattcui Actually the unit test [fs_agent_env_service_test.go](https://github.com/cloudfoundry/bosh-softlayer-cpi-release/blob/master/src/bosh-softlayer-cpi/softlayer/common/fs_agent_env_service_test.go#L111-L122) is already removed in [my fork](https://github.com/vvraskin/bosh-softlayer-cpi/blob/0fddfc4e628fa1ae3591dc68e22ec320dbf1ef06/src/bosh-softlayer-cpi/softlayer/common/fs_agent_env_service_test.go#L111) and in this PR. [LengthOfHostName](https://github.com/vvraskin/bosh-softlayer-cpi/blob/0fddfc4e628fa1ae3591dc68e22ec320dbf1ef06/src/bosh-softlayer-cpi/action/create_vm.go#L101) is required to verify the length in this [UT](https://github.com/vvraskin/bosh-softlayer-cpi/blob/0fddfc4e628fa1ae3591dc68e22ec320dbf1ef06/src/bosh-softlayer-cpi/action/create_vm_test.go#L134). Do you still think that we need to remove the variable `LengthOfHostName`?

mattcui commented:
@vvraskin Ok, I would like to leave the decision to @maximilien.

@maximilien Actually, the public variable [LengthOfHostName](https://github.com/cloudfoundry/bosh-softlayer-cpi-release/blob/master/src/bosh-softlayer-cpi/softlayer/common/helper/softlayer_helper.go#L20)  is not needed, but it would be useful for [this UT](https://github.com/vvraskin/bosh-softlayer-cpi/blob/0fddfc4e628fa1ae3591dc68e22ec320dbf1ef06/src/bosh-softlayer-cpi/action/create_vm_test.go#L108-L139), do you think if we could remove this variable and the UT as well? Thanks.

maximilien commented:
I agree with @mattcui as by having this var public could create a security risk plus allow the value to change by any code (so hard to track).

@vvraskin any way to test and making this public or make it a const (perhaps)?

vvraskin commented:
@mattcui @maximilien Ok, agreed. Please see the latest commit. I've got rid of the usage of `LengthOfHostName` variable coming from the helper, added a new public method [GetVMCloudProperties](https://github.com/vvraskin/bosh-softlayer-cpi/blob/master/src/bosh-softlayer-cpi/action/create_vm.go#L35-L37) where I get the length from and updated the unit tests accordingly. Please review. 

regards, Vadim

maximilien commented:
@mattcui and @jianqiu I am adding you both as reviewers. Please take a look at let's discuss in scrum tonight and get back with @vvraskin. My main concern is what impact on our tests envs and Bluemix team.

mattcui commented:
@maximilien The only potential problem I can see is there will be some trouble when the length of director hostname is 64, but I think we just need to document it to ask the user to not set 64 characters hostname when creating director, @gu-bin will address it. So in my opinion, you can merge this PR now. Thanks.

maximilien commented:
Thanks @mattcui. Merging.

/cc @gu-bin please ensure you have docs updated before next release. Thx

zhanggbj commented:
@vvraskin, please note this PR is reverted as we talked in Slack.

An error occurs when calling create_vm action:
```
STDOUT: '{"result":null,"error":{"type":"Bosh::Clouds::CloudError","message":"Run method not found","ok_to_retry":false},"log":""}'
```

Please fix it and resubmit your PR. Thanks!

vvraskin commented:
@zhanggbj When I submitted the PR all the tests were green, do you think that it would be better if you create a test which will fail under the same PR? And then I will fix it.

zhanggbj commented:
@vvraskin as we just talked on slack, I'll take a look to improve the dispatcher UT to cover this case, but for this PR, maybe changing another way to implement create_vm UT is a better and faster option. Thanks.

mattcui commented:
@zhanggbj  I think you could show @vvraskin on how to run create_vm integration test to recreate the problem.

zhanggbj commented:
Sure @mattcui @vvraskin , here are steps to reproduce it. Thanks.

- Build softlayer_cpi binary under src/bosh-softlayer-cpi/
```
go build -o out/softlayer_cpi_debug main/main.go
```

- Prepare cpi.json 
```
{
    "cloud": {
        "plugin": "softlayer",
        "properties": {
            "softlayer": {
                "username": "fake-username",
                "apiKey": "fake-api-key",
                "featureOptions": {}
            },
            "agent": {
                "ntp": [
                    "0.pool.ntp.org",
                    "1.pool.ntp.org"
                ],
                "blobstore": {
                    "provider": "dav",
                    "options": {
                        "endpoint": "http://127.0.0.1:25250",
                        "user": "agent",
                        "password": "fake-password"
                    }
                },
                "mbus": "nats://nats:{fake-password}@director-v2.softlayer.com:4222"
            },
            "baremetal": {},
            "pool": {}
        }
    }
}
```
- Run create_vm integration test as below
```
out/softlayer_cpi -configPath cpi.json < dev/create_vm.json
```

maximilien commented:
All, FYI, for building CPI, the `src/bosh-softlayer-cpi/bin/` should also be a usable option. And perhaps preferred since that's what the pipeline should use to create binaries.
